Task: Evaluate & Clean up Data from an E-Commerce Company

## Thinking & Targeting

The goal is to find popular products based on sales data, for later strategy making and enhancing business performance.

This project is aimed to practise evaluating structure and content of the data, and then cleans the data based on the findings to make data ready for analysis.

## Introduction

The global sales data was originally collected by an E-Commerce Company between 1st, Dec.2010 and 9th, Dec. 2011. 
The company mainly sells gifts covering various scenarios, including but not limited to birthday gifts, wedding souvenirs, Christmas gifts, and so on. The company's customer base mainly consists of wholesalers and individual consumers, among whom wholesalers account for a considerable proportion.

Columns in the table：
- `InvoiceNo`: Invoice number. 6 digits, serving as a unique identifier for the transaction. If this code starts with the letter 'c', it indicates that the transaction has been cancelled.
- `StockCode`: Product code. A five-digit number serves as the unique identifier for the product.
- `Description`: product name.
- `Quantity`: The quantity of the product in the transaction.
- `InvoiceDate`: Invoice date and time. The date and time when the transaction occurred.
- `UnitPrice`: Unit price. The price is stated in pounds (£).
- `CustomerID`: Customer number. A 5-digit number that serves as the unique identifier for the customer.
- `Country`: Country name. The name of the country where the customer resides.

## Data Loading

Import NumPy, Pandas，use 'read_csv' to read'e_commerce.csv'，Convert it to DataFrame format and assign it to 'df'.

Display the first 5 rows and the last 5 rows of the dataset.

In [2]:
import numpy as np
import pandas as pd

In [6]:
df= pd.read_csv('e_commerce.csv')
df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


## Data Evaluating

In this section, we evaluate  the data contained in the DataFrame that was created in the previous section.

The evaluation is conducted from two main aspects: structure and content, namely, the organization and cleanliness of the data. Structural issues refer to violations of the three criteria: “each column represents a variable, each row represents an observation, and each cell contains a value.” Content-related issues include missing data, duplicate data, and invalid data.

### Evaluate data cleanliness

Using the info() and sample() methods of the DataFrame, we can obtain an overview of the data as well as sample data itself, thereby assessing the cleanliness of the data.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Column names comply with the rule that each column should represent only one variable.

In [7]:
df.sample(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
52469,540798,21027,NINJA RABBIT PINK,2,1/11/2011 12:11,1.25,17365.0,United Kingdom
105364,545221,20979,36 PENCILS TUBE RED RETROSPOT,16,3/1/2011 8:35,1.25,14740.0,United Kingdom
405482,571714,22499,WOODEN UNION JACK BUNTING,1,10/18/2011 17:00,12.46,NaN,United Kingdom
63192,541524,22432,WATERING CAN PINK BUNNY,6,1/19/2011 10:49,1.95,17091.0,United Kingdom
400780,571372,21744,SNOWFLAKE PORTABLE TABLE LIGHT,2,10/17/2011 11:55,2.95,17201.0,United Kingdom


Based on the 5 rows of data sampled, they meet all three criteria: each column represents a variable, each row represents an observation, and each cell contains a single value. Specifically, each row contains transaction data for a particular product, and each column represents various aspects of that transaction. Therefore, there are no structural issues with the data.

### Evaluate data cleanliness

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


Based on the output results, there are a total of 541,909 observations in the dataset.
There are missing values in the `Description` and `CustomerID` columns.
The data type of the `CustomerID` column should be “object”, and the data type of the `InvoiceDate` column should be “datetime”. Data type conversions are required.

#### Evaluate missing values

Preliminary count of missing values in each column

In [9]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

After determining that there are missing values in the `Description` field, the missing observations were extracted based on the specified conditions.

In [10]:
df[df.Description.isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


1454 transaction records lack the `Description` variable value.

From the output results, it can be seen that for these transaction records lacking a “Description”, the “UnitPrice” is 0, and the “CustomerID” is also missing. 
To verify this hypothesis, we added additional filtering criteria to check if there are any records where the “Description” is missing but the “UnitPrice” is not 0.

In [11]:
df[(df.Description.isnull()) & (df.UnitPrice != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


Verifying the hypothesis: The output shows 0 entries, indicating that there is missing data for the `Description` variable, and thus no valid data available.

`Description` represents the product name, and `UnitPrice` represents the unit price of the product. Both are key variables for subsequent data analysis. If both are missing, we consider the data to be meaningless, so such records can be deleted later.

The variable `CustomerID` also contains missing values. Therefore, missing observations were extracted based on the specified conditions.

In [12]:
df[df.CustomerID.isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


135,080 transaction records lack the `CustomerID` variable value.

Based on the output results, the absence of the `CustomerID` variable does not necessarily mean that other variables are also missing. Additionally, `CustomerID` isn’t a key variable for analyzing best-selling products. Therefore, observations with empty `CustomerID` values will be retained as is.

#### Evaluate the number of repetitions

Based on the meaning of these data fields, although `InvoiceNo`, `StockCode`, and `CustomerID` are all unique identifiers, a single transaction can involve multiple items. Therefore, `InvoiceNo` can appear multiple times. Additionally, the same item can be selected across different transactions, so `InvoiceNo` can still be repeated. Each customer can initiate multiple transactions, which means `CustomerID` can also appear multiple times.

However, when all variables in the data are duplicated at the same time, it should be considered duplicate data. It’s not the case that a customer buys multiple copies of the same product at once – in that case, the value of the “quantity” variable would be multiple.

In [13]:
df[df.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,12/1/2010 11:45,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,12/1/2010 11:45,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,12/1/2010 11:45,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,12/1/2010 11:45,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,12/1/2010 11:49,2.95,17920.0,United Kingdom
...,...,...,...,...,...,...,...,...
541675,581538,22068,BLACK PIRATE TREASURE CHEST,1,12/9/2011 11:34,0.39,14446.0,United Kingdom
541689,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,12/9/2011 11:34,2.49,14446.0,United Kingdom
541692,581538,22992,REVOLVER WOODEN RULER,1,12/9/2011 11:34,1.95,14446.0,United Kingdom
541699,581538,22694,WICKER STAR,1,12/9/2011 11:34,2.10,14446.0,United Kingdom


Based on the output results, there are 5268 duplicate entries in the data. All duplicate entries across all variables should be removed during the data cleaning process.

#### Evaluating inconsistent data 

Evaluating inconsistent data from the data variables.

In [14]:
df.Country.value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

Based on the values in the `Country` variable, “U.K.”, “UK”, and “United Kingdom” all refer to the United Kingdom, while “United States” and “USA” both refer to the United States. These values are inconsistent. In subsequent cleaning steps, it’s necessary to replace all these values with a unified representation.

#### Evaluate invalid/error data

Use the `describe` method of the DataFrame to quickly obtain numerical statistical information.

In [15]:
df.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


Based on the output results, there are negative values for both `Quantity` and `UnitPrice`. These values are invalid.

Next, let’s filter out the observations with a negative `Quantity` value, in order to better understand the characteristics of the data.

In [16]:
df[df['Quantity'] <= 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


It can be observed that when `Quantity` is negative, it seems that `InvoiceNo` always starts with “C”, indicating that the transaction was cancelled.

To verify this hypothesis, we need to add additional filtering criteria to check if there are any observations where the `Quantity` is negative and the `InvoiceNo` does not start with “C”.

In [17]:
df[(df['Quantity'] <= 0) & (df['InvoiceNo'].str[0] != 'C')]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


It can be seen that the hypothesis is not accurate. There indeed are observations where the Quantity variable is negative and the InvoiceNo does not start with “C”.

However, upon further inspection, it was found that the UnitPrice value for all these selected observations was 0. Therefore, we added additional filtering criteria to verify this further.

In [18]:
df[(df['Quantity'] <= 0) & (df['InvoiceNo'].str[0] != 'C') & (df['UnitPrice'] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


Based on the output，The hypothesis is confirmed. It turns out that when `Quantity` is negative, the observed values satisfy one of the following conditions:

`InvoiceNo` starts with “C”, indicating that the order has been cancelled.

`UnitPrice` is 0. This indicates that the unit price is 0 pounds.

These transaction data are not valid transaction records and are therefore not included in the subsequent data analysis. As such, during the data cleaning phase, observations with a negative `Quantity` value were removed.

At the same time, we need to confirm whether the `Quantity` variable is always negative when `InvoiceNo` starts with “C” or when `UnitPrice` is 0.

This will help us determine whether, during the data cleaning process, we only need to delete records with a negative `Quantity` value.

In [19]:
df[(df['InvoiceNo'].str[0] == 'C') & (df['Quantity'] > 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


According to the output results, when `InvoiceNo` starts with “C”, the value of the `Quantity` variable is always greater than 0. Therefore, during the data cleaning phase, by removing any observations where `Quantity` is negative, all records of cancelled orders can be eliminated.

Because when evaluating the missing data, it was found that there were entries where the `Description` field was empty, and the `UnitPrice` was 0. Therefore, a condition must be added during filtering: the `Description` field cannot be empty.

In [20]:
df[(df['UnitPrice'] == 0) & (df['Quantity'] > 0) & (df['Description'].notnull())]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,12/3/2010 12:08,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,12/3/2010 12:08,0.0,NaN,United Kingdom
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,12/5/2010 14:02,0.0,12647.0,Germany
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,12/7/2010 11:48,0.0,NaN,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,12/7/2010 11:48,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
524622,580609,22927,Amazon,1,12/5/2011 11:41,0.0,NaN,United Kingdom
535325,581202,23404,check,41,12/7/2011 18:30,0.0,NaN,United Kingdom
535334,581211,22142,check,14,12/7/2011 18:36,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,12/8/2011 13:58,0.0,NaN,United Kingdom


According to the output results, there are instances where the `UnitPrice` is 0, the `Quantity` is greater than 0, and the `Description` is not empty.

In [21]:
df[(df['UnitPrice'] == 0) & (df['Quantity'] > 0) & (df['Description'].notnull()) & (df['CustomerID'].notnull())]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,12/5/2010 14:02,0.0,12647.0,Germany
33576,539263,22580,ADVENT CALENDAR GINGHAM SACK,4,12/16/2010 14:36,0.0,16560.0,United Kingdom
40089,539722,22423,REGENCY CAKESTAND 3 TIER,10,12/21/2010 13:45,0.0,14911.0,EIRE
47068,540372,22090,PAPER BUNTING RETROSPOT,24,1/6/2011 16:41,0.0,13081.0,United Kingdom
47070,540372,22553,PLASTERS IN TIN SKULLS,24,1/6/2011 16:41,0.0,13081.0,United Kingdom
56674,541109,22168,ORGANISER WOOD ANTIQUE WHITE,1,1/13/2011 15:10,0.0,15107.0,United Kingdom
86789,543599,84535B,FAIRY CAKES NOTEBOOK A6 SIZE,16,2/10/2011 13:08,0.0,17560.0,United Kingdom
130188,547417,22062,CERAMIC BOWL WITH LOVE HEART DESIGN,36,3/23/2011 10:25,0.0,13239.0,United Kingdom
139453,548318,22055,MINI CAKE STAND HANGING STRAWBERY,5,3/30/2011 12:45,0.0,13113.0,United Kingdom
145208,548871,22162,HEART GARLAND RUSTIC PADDED,2,4/4/2011 14:42,0.0,14410.0,United Kingdom


Based on the output results, the hypothesis is confirmed. It turns out that when `UnitPrice` is 0, for records where `Quantity` is greater than 0 and `Description` is not empty, most `CustomerID` values are empty, while a small portion are not.

However, since the ultimate goal of data analysis is to increase revenue, data with a `UnitPrice` of 0 is deleted during the data cleaning process.

Next, filter out the observations with a negative `UnitPrice` to examine the characteristics of the data.

In [22]:
df[df['UnitPrice'] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


Based on the output results, all observations where `UnitPrice` is negative represent bad debt adjustments and do not represent actual transaction data. Therefore, they are also removed during the data cleaning process.

Using the `value_counts` method of the DataFrame, check whether any invalid/error data can be detected in the `InvoiceNo`, `StockCode`, and `InvoiceDate` variables.

In [23]:
df.InvoiceNo.value_counts()

573585     1114
581219      749
581492      731
580729      721
558475      705
           ... 
558482        1
542549        1
560218        1
C577164       1
546148        1
Name: InvoiceNo, Length: 25900, dtype: int64

In [24]:
df.StockCode.value_counts()

85123A    2313
22423     2203
85099B    2159
47566     1727
20725     1639
          ... 
84966A       1
35951        1
84876C       1
21330        1
35597B       1
Name: StockCode, Length: 4070, dtype: int64

In [25]:
df.InvoiceDate.value_counts()

10/31/2011 14:41    1114
12/8/2011 9:28       749
12/9/2011 10:03      731
12/5/2011 17:24      721
6/29/2011 15:58      705
                    ... 
3/28/2011 12:02        1
9/2/2011 12:41         1
6/2/2011 11:51         1
7/14/2011 14:55        1
1/27/2011 14:26        1
Name: InvoiceDate, Length: 23260, dtype: int64

Based on the output results, no invalid or incorrect data was detected.

## Clean up data

Based on the results of the data evaluation section above, the data cleaning tasks that need to be performed include:

- Convert the data type of the `CustomerID` variable to string.
- Convert the data type of the `InvoiceDate` variable to datetime.
- Delete any missing observations for the `Description` variable
- Delete all duplicate data for all variables
- Replace the values of the `Country` variable, namely `'U.K.'` and `'UK'`, with `'United Kingdom'`
- Replace the value of the `Country` variable from `'USA'` to `'United States'`
- Delete observations where `Quantity` is less than or equal to 0.
- Delete data where the `UnitPrice` value is less than or equal to 0.

To distinguish between the cleaned data and the original data, we create a new variable called `cleaned_df`. All subsequent cleaning steps will be applied to `cleaned_df`.

In [26]:
cleaned_df = df
cleaned_df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France


Convert the data type of the `CustomerID` variable to string:

In [27]:
cleaned_df['CustomerID'] = cleaned_df['CustomerID'].astype(str)
cleaned_df['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

Remove “.0” from the end of `CustomerID`:

In [28]:
cleaned_df['CustomerID'] = cleaned_df['CustomerID'].str.slice(0, -2)
cleaned_df['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

Convert the data type of the `InvoiceDate` variable to datetime type:

In [31]:
cleaned_df['InvoiceDate'] = pd.to_datetime(cleaned_df['InvoiceDate'])
cleaned_df.InvoiceDate

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

Delete observations with missing values for the `Description` variable:

In [32]:
cleaned_df.dropna(subset=['Description'], inplace=True)
cleaned_df.Description.isnull().sum()

0

Delete all duplicate data across all variables：

In [33]:
cleaned_df.drop_duplicates(inplace=True)
cleaned_df[cleaned_df.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


Replace the values of the `Country` variable, namely `'U.K.'` and `'UK'`, with `'United Kingdom'`.
Replace the value of the `Country` variable from `'USA'` to `'United States'`:

In [34]:
cleaned_df.replace(['U.K.', 'UK'], 'United Kingdom', inplace=True)
cleaned_df.replace('USA', 'United States', inplace=True)
print(len(cleaned_df[cleaned_df['Country'] == 'U.K.']))
print(len(cleaned_df[cleaned_df['Country'] == 'UK']))
print(len(cleaned_df[cleaned_df['Country'] == 'USA']))

0
0
0


Delete observations where `Quantity` is less than or equal to 0.

In [35]:
cleaned_df.drop(cleaned_df[cleaned_df.Quantity <= 0].index, inplace=True)
len(cleaned_df[cleaned_df.Quantity <= 0])

0

Delete data where the `UnitPrice` value is less than or equal to 0:

In [36]:
cleaned_df.drop(cleaned_df[cleaned_df.UnitPrice <= 0].index, inplace=True)
len(cleaned_df[cleaned_df.UnitPrice <= 0])

0

It was found that the unit price included 0.001 pounds, which is unreasonable and likely incorrect data.

In [37]:
cleaned_df.drop(cleaned_df[cleaned_df.UnitPrice == 0.001].index, inplace=True)
cleaned_df.UnitPrice.sort_values()

301033        0.04
221726        0.04
363674        0.04
410140        0.04
407634        0.04
            ...   
268028     4287.63
297723     4505.17
173382     8142.75
299982    11062.06
15017     13541.33
Name: UnitPrice, Length: 524874, dtype: float64

## Save the cleaned data

After completing the data cleaning process, save the cleaned and organized data in a new file named `e_commerce_cleaned.csv`.

In [38]:
cleaned_df.to_csv('e_commerce_cleaned.csv', index=False)
pd.read_csv('e_commerce_cleaned.csv')

C:\Users\xthau\miniconda3\envs\py36\lib\site-packages\IPython\core\interactiveshell.py:3072: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
...,...,...,...,...,...,...,...,...
524869,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680,France
524870,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680,France
524871,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680,France
524872,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680,France


When df_cleaned is output, there are warning messages, which is consistent with the reference version.